src/ORM/ResultSet.php

```diff
@@ -156,25 +140,23 @@ class ResultSet implements ResultSetInterface
      * Constructor
      *
      * @param \Cake\ORM\Query $query Query from where results come
-     * @param \Cake\Database\StatementInterface $statement The statement to fetch from
+     * @param array $results Results array.
      */
-    public function __construct(Query $query, StatementInterface $statement)
+    public function __construct(Query $query, array $results)
     {
         $repository = $query->getRepository();
-        $this->_statement = $statement;
         $this->_driver = $query->getConnection()->getDriver();
         $this->_defaultTable = $repository;
         $this->_calculateAssociationMap($query);
         $this->_hydrate = $query->isHydrationEnabled();
         $this->_entityClass = $repository->getEntityClass();
-        $this->_useBuffering = $query->isBufferedResultsEnabled();
         $this->_defaultAlias = $this->_defaultTable->getAlias();
         $this->_calculateColumnMap($query);
         $this->_autoFields = $query->isAutoFieldsEnabled();
 
-        if ($this->_useBuffering) {
-            $count = $this->count();
-            $this->_results = new SplFixedArray($count);
+        $this->__unserialize($results);
+        foreach ($this->_results as $i => $row) {
+            $this->_results[$i] = $this->_groupResult($row);
```

Comment:
We should add a comment above explaining why this is required after unserialize.

Reply:
Well the call to `__serialize()` is just to avoid code duplication. Also the `_groupResult()`'s doc states what it does.